lib/sqlalchemy/orm/attributes.py

```diff
@@ -270,6 +270,13 @@ class InstrumentedAttribute(QueryableAttribute):
     """
 
     def __set__(self, instance, value):
+        if self.impl.readonly:
+            msg = (
+                "the value set to read only attribute '%s.%s' will be ignored."
+                " A future version of sqlalchemy will change this warning to "
+                "raise an exception"
+            ) % (self.class_.__name__, self.key)
+            util.warn(msg)
         self.impl.set(
```

Comment:
one more change that I can make if you want me to do it on the gerrit side, the warning here could come from the set() method of the impl, that way there's no change here and fewer attribute lookups

Reply:
I originally added it here because I don't know  of `impl`can be an instance of different classes, so moving it in the `impl.set` may duplicate the code.

I'm not sure if an attribute lookup will change much in the grand scheme of things. I guess we could do

```py
impl = self.impl
if impl.readlony:
  ...
impl.set()
```

So we are doing the same number of lookups as if the code was moved in` set`

Reply:
OK, so, here is something, this behavior actually only applies to the ScalarAttributeImpl.    I think we might be better off for now making it local to that.   It definitely doesn't apply correctly to CollectionAttributeImpl since we aren't blocking collection operations.    then we can set "readonly" only in the case that we are dealing with a column loader, like this:

```diff
diff --git a/lib/sqlalchemy/orm/strategies.py b/lib/sqlalchemy/orm/strategies.py
index 79e36026c..8fc8a63a8 100644
--- a/lib/sqlalchemy/orm/strategies.py
+++ b/lib/sqlalchemy/orm/strategies.py
@@ -207,6 +207,7 @@ class ColumnLoader(LoaderStrategy):
             useobject=False,
             compare_function=coltype.compare_values,
             active_history=active_history,
+            readonly=self in mapper._readonly_props
         )
 
     def create_row_processor(
@@ -283,6 +284,7 @@ class ExpressionColumnLoader(ColumnLoader):
             useobject=False,
             compare_function=self.columns[0].type.compare_values,
             accepts_scalar_loader=False,
+            readonly=True   #  because this is ExpressionColumnLoader
         )
 
 
@@ -342,6 +344,7 @@ class DeferredColumnLoader(LoaderStrategy):
             compare_function=self.columns[0].type.compare_values,
             callable_=self._load_for_state,
             load_on_unexpire=False,
+            readonly=self in mapper._readonly_props
         )
 
     def setup_query(

```

sorry this keeps changing but I think that would be better.  As always I am willing to pick this up since I have led you around a bunch already :)

Reply:
Please feel free to continue this. I'm really not familiar with this part, and is not very easy to follow the code, so at this point I'm just guessing.

Reply:
OK, no problem, thanks for your efforts on this!